### DISCUSSION TITLE
Saturation/Distortion options

### DISCUSSION THREAD

**Discussion by manysounds** [SynthstromAudible/DelugeFirmware#353] [Ideas]
Some other distortion/saturation mode/modes other than the default saturation would be immensely appreciated. 
Alternatively added under the spot held by the new in-progress wavefolder, so row 2 column 8.
How to switch between multiple algorithms I'm not sure. Encoder press seems most obvious or r2c8 press again.
Gloubi-boulga, Tanh, soft-clip, hard-clip, some diode-clip emulation, old tape/vinyl?
Some of these things can be a bit processor intensive and use lookup tables for frequency dependent clipping (i.e. Volterra) so we can just skip that... for now maybe. Some may also require additional parameters, like old tape emulation saturation/wobble/flutter.
A cheap soft-clip/LPF combo can sound wonderful for drums and all kinds of other things.
Big dreams :)

**Comment by trappar** [comment]
Dirtywave M8 has quite a few distortion/saturation algos and it runs on hardware far less powerful than deluge, so I feel like a lot is possible!

**Reply by m-m-adams** [reply]
The M8 processor is in the same ballpark as the deluge and it's running less than a quarter of the voice count, but yes there are distortion options that aren't too hard on DSP

**Reply by trappar** [reply]
M8's max polyphony right now is actually 104 voices if you use 8 instances of hypersynth. Each track has distortion that I assume is applied on an already summed signal, so a maximum of 8 stereo distortion instances.

This seems fairly similar to deluge with the exception of kits where each row gets its own distortion.

Maybe a reasonable limitation would be to add some slightly more processor intensive distortion options which only work when affect entire is enabled?

**Reply by m-m-adams** [reply]
Still by the same token deluge can hit 4-500 voices if you count unison

**Comment by m-m-adams** [comment]
Would you mind checking out the latest nightlies? PR #336 added drive to the ladder filters (labelled morph). It's in the menu or on the pad under db/Oct. I am contemplating moving filter drive into its own pad and then putting filter FM in the morph pad instead, you can test the FM with the high pass ladder in #339 for comparison 

Separately if you could test #349 that would be appreciated - currently on the pad between saturation and low pass frequency, I'm planning to add a mode switch with other distortion options as well

**Reply by manysounds** [reply]
I don't have building set up locally here, apologies. I'm working off the pull request builds.

**Reply by manysounds** [reply]
OK, I've tested those builds and they're not what I've been hoping for.
The drive filter is definitely useful for some synth sounds, which is great.
The Wavefolder seems mostly useless except at extreme settings and does not work well for drum tracks.
I feel that we could get the most use out of alternate distortion/overdrive algorithms for synths/drum kits/songs/affect entire etc so we can distort/overdrive separately from the filter. The drive filter doesn't show it's colors until the cutoff frequency becomes apparent which is a less flexible approach.
Perhaps multiple presses of Shift-Saturation (or shift-wavefolder) could cycle between different distortion/wavefolding algorithms.
If I had a clue how to program in this environment I'd be all over it, sorry for my lack of experience in this area.

**Reply by legsmechanical** [reply]
Just want to make sure the wavefolder isn’t being devalued. I think it’s a great tool, especially when automated in the sweet spots (which will vary depending on what you’re sending it. )

**Reply by manysounds** [reply]
No no, I don’t think it should be removed at all.

**Reply by m-m-adams** [reply]
The wave folder effect depends heavily on what you're putting in to it - try with a sine wave and all the settings are useful, conversely with a square it'll change the volume as much as it works as an effect

The filter drive isn't limited to the drive filter anymore - the low pass ladder filters are all boostable with the pad underneath the db/Oct (on the SVF it's a morph and HPF its currently filter FM). On the drive filter it roughly doubles the available gain 

I am planning to add more waveshaping/distortion in the future though

**Comment by manysounds** [comment]
Distortion/Overdrive/Clipping options separate from filter would be preferred

**Reply by trappar** [reply]
Maybe there could be options for setting whether distortion is applied pre or post filter?

Either way I agree it should be separate.

**Reply by manysounds** [reply]
Per instrument and per song, yes
Some anti-aliasing wiggles may be needed, just thinking

**Reply by m-m-adams** [reply]
The filter drive is internal to the filter (it's on the feedback path) and can't be moved. I think I can expand the wavefolding into waveshaping and add some more traditional distortion options though!

**Comment by manysounds** [comment]
Simple follow up.
Latest nightly 2023_09_29-c56cdb5.
The filter drive reacts interestingly. On a setting of 2 it completely crushes and overcompresses wildly. At higher settings it's driven but the volume drops precipitously. Not sure what we were going for here but it seems backwards at first, and then proper, then not right.
What i'd TRULY love to see is an overdrive/clipper in the Wavefold slot that works on "affect entire" for drum kits.  Simple sine loving wavefolding is not going to cut butter. 
Dumping a drum machine through a Big Muff (for example) is a sought after sound. Playing samples (drum loops or otherwise) through a some kind of distortion and THEN using a bandpass is a lo-fi classic.
I appreciate the new filter options but it's not what I was looking for at the start of this thread. 
A simple **"x = atan(xin * drive)"** maybe 
A rewrite of the **Max/MSP gen~ codebox "softclip"?**
scSoftclip(xin, drive) {
	drive = maximum(drive, 0.);
	x1 = xin * drive;
	x2 = (x1 * x1) + 0.25;
	return x1 / x2;
}
or the venerated **Gloubi-Bougla:**
const double x = input * 0.686306;
const double a = 1 + exp (sqrt (fabs (x)) * -0.75);
output = (exp (x) - exp (-x * a)) / (exp (x) + exp (-x));

I am entirely unaware of what system DSP we are leaving available for this effect slot but run a drum loop through those and be excited! Also, making these available for Affect Entire for kits!
My experience is in Max/MSP, a little Superconductor, and some other not-a-real-programmer-things so I'm of little real coding help, apologies.

**Comment by aureloctopuzzle** [comment]
I have a point to make about the distortion: it’s "on/off" right from the very first setting—it’s already too much!
